resteasy-jaxrs/src/main/java/org/jboss/resteasy/core/StringParameterInjector.java

```diff
@@ -80,40 +80,39 @@ protected void initialize(Class type, Type genericType, String paramName, Class
       baseType = type;
       baseGenericType = genericType;
 
-      if (type.isArray()) baseType = type.getComponentType();
-      if (List.class.isAssignableFrom(type))
-      {
-         isCollection = true;
-         collectionType = ArrayList.class;
-      }
-      else if (SortedSet.class.isAssignableFrom(type))
-      {
-         isCollection = true;
-         collectionType = TreeSet.class;
-      }
-      else if (Set.class.isAssignableFrom(type))
+      if (type.isArray())
       {
-         isCollection = true;
-         collectionType = HashSet.class;
+    	  baseType = type.getComponentType();
       }
-      if (isCollection)
+     
+      if (!baseType.isPrimitive())
       {
-         if (genericType != null && genericType instanceof ParameterizedType)
+         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
+         if (paramConverter != null) 
          {
-            ParameterizedType zType = (ParameterizedType) genericType;
-            baseType = Types.getRawType(zType.getActualTypeArguments()[0]);
-            baseGenericType = zType.getActualTypeArguments()[0];
+        	 return;
          }
-         else
-         {
-            baseType = String.class;
-            baseGenericType = null;
+         
+         collectionType = convertParameterTypeToCollectionType();
+         if(collectionType != null){
+        	 isCollection = true;
+	         if (genericType != null && genericType instanceof ParameterizedType)
+	         {
+	            ParameterizedType zType = (ParameterizedType) baseGenericType;
+	            baseType = Types.getRawType(zType.getActualTypeArguments()[0]);
+	            baseGenericType = zType.getActualTypeArguments()[0];
+	         }
+	         else
+	         {
+	            baseType = String.class;
+	            baseGenericType = null;
+	         }
+	         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
+	         if (paramConverter != null) 
+	         {
+	        	 return;
+	         }
          }
-      }
-      if (!baseType.isPrimitive())
-      {
-         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
-         if (paramConverter != null) return;
 
          unmarshaller = factory.createStringParameterUnmarshaller(baseType);

```

Comment:
re: "It will allow users to provide a way to convert String type to any type subclassing List, Set or SortedSet before fallback on default implementation ArrayList, TreeSet and HashSet"

So, use the actual type found in the resource method. Hmmm, I wonder why the type is getting "normalized" now ...